core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeTransformationManager.java

```diff
@@ -104,7 +120,45 @@ public void run() {
                 , 1, 1, TimeUnit.SECONDS);
     }
 
-    private static String createDeviceTransformPayload(List<Integer> rowIds, List<String> messages) {
+    private void completeDeviceModeEventProcessing() {
+        RudderLogger.logDebug(String.format(Locale.US, "DeviceModeTransformationManager: TransformationProcessor: Updating status as DEVICE_MODE_PROCESSING DONE for events %s", messageIds));
+        dbManager.markDeviceModeDone(messageIds);
+        dbManager.runGcForEvents();
+    }
+
+    RudderMessage getEventFromMessageId(int messageId) {
+        String message = messages.get(messageIds.indexOf(messageId));
+        return gson.fromJson(message, RudderMessage.class);
+    }
+
+    // For each message get the list of destinationIds for which transformation is enabled
+    private ArrayList<String> getTransformationEnabledDestinationIds(List<String> messages) {
+        ArrayList<List<String>> destinationIds = new ArrayList<>();
+        for (int i = 0; i < messages.size(); i++) {
+            RudderMessage message = gson.fromJson(messages.get(i), RudderMessage.class);
+            destinationIds.add(this.rudderDeviceModeManager.getTransformationEnabledDestinationIds(message));
+        }
+        return convertIntoString(destinationIds);
+    }
+
+    private ArrayList<String> convertIntoString(List<List<String>> destinationIds) {
```

Comment:
Why can't we propagate this and let gson do the serialization?

Reply:
We are manually serialising the request-response over [here](https://github.com/rudderlabs/rudder-sdk-android/blob/f4dbcfa3cc626db146ae39fe14b7337448e4ab67/core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeTransformationManager.java#L161), hence I did it in the same way. 

Please LMK, if there is a better way to do this.